### DISCUSSION TITLE
The future of Piranha

### DISCUSSION THREAD

**Discussion by tidyui** [PiranhaCMS/piranha.core#1924] [Ideas]
In order to plan for the future, in terms of the next major versions, and even the next generation of Piranha CMS, it would be awesome if you could write a short reply with:

1. What are the key features you use Piranha CMS for?
2. What are the key features you would like to see in the future to make it even more useful?

Best regards!

**Comment by jensbrak** [comment]
Since I don't use Piranha in a commercial or even a production like setting I don't want to risk influencing too much. Just adding my thoughts to get this discusssion started (hopefully). 

(Background: I use Piranha CMS mostly for the fun of it. Found it while trying to learn more about modern ASP.NET web sites and everything in the eco system for those - from the practical aspects of C# patterns and best practices various more top level implementations of common scenarios - like user management and hosting challenges. I have no actual use case for Piranha, yet.)

1. Piranha CMS provides an excellent platform to get a "head start" on an actual, useful, content management platform (rather than starting from empty or near empty MS templates). It also provides the basic structure of a useful CMS where the most important building blocks are ready made, but all made to be easily extended and modified.

So for me personally I use Piranha CMS as an educational tool. The key features I've found liberating to use is the Manager with everything that comes with it, as well as the Core platform providing content management. However, the key feature of Piranha must be that it is open source, with active, skilled and helpful project members and community members alike. 

So not really what was asked for in terms if feature X or Y, but still relevant I hope.

2. Bare in mind the context in which I use Piranha. I'd say the best thing to do would be more up to date examples and documentation. Really. Since it is a platform that can be extended - the shorter and easier the path to extend and contribute is - the more likely it is that the useful but missing features will be added by the community rather than the core dev team. Or at least that is what I would hope.

To name a couple of things I myself found missing or would love to get a heads start on in Piranha: some sort of sample implementation of a central authentication manager, for instans "log in using Github". I mean, it's almost standard for any site to provoide such a functionality if user login is possible. Yet it is surprisingly complicated to get these things running. I'v tried (somewhat half hearted).


Just my five cents, but then again. I am no "real" user of Piranha.

**Reply by GerardoGala** [reply]
I agree. Authentication and login page should have come out of the box.

**Comment by CarlSargunar** [comment]
I've just discovered Piranha while listening to a podcast, and they were discussing Open Source .NET CMS systems. I'm from an primarily Umbraco background, but I love the idea of a super lightweight CMS, which is EVEN lighter than Umbraco, and is completely open source. 

I'm not at all very experienced with Piranha, but I hope to spend some time migrating my blog over to use the CMS as a proof of concept, and maybe will continue to use it for smaller (or larger) sites going forward

**Reply by tidyui** [reply]
That’s cool, just reach out if you have any questions! Do you have a link to the podcast btw, sounds like an interesting listen 🙂

**Reply by CarlSargunar** [reply]
It was mentioned in this one : https://dotnetcore.show/episode-106-fim-market-and-the-net-tech-stack-with-miguel-adwin/

**Comment by debeardjr** [comment]
I have Piranha integrated into a .NET 6 MVC multi-tenant application and it works seamlessly. I absolutely love the API and ability to extend the blocks you have created. There are up to three pages per site that are CMS driven. The rest is controlled, data driven, by the MVC application. 

We have internal users that belong to a content management group and that group is given permission to access the Piranha manager UI. Our content managers are responsible for certain tenants but currently we have to expose all sites/tenants in the Piranha manager for all content managers. It would be very helpful if the Piranha manager had the ability to link a web site to a custom security claim so that we can control access to web sites. This would allow us to expose only sites/tenants that a content manager is responsible for.

Thanks for this awesome product!

**Reply by tidyui** [reply]
We're looking into adding hooks for the manager models just like for the application models. This way you could for example hook into the `OnLoad` hook for `PageList` and simply filter the model depending on the current user (or anything else you wanted for that matter).

**Comment by EddyCwdry** [comment]
I use Piranha extensively for multiple customer sites at our company, we provide fintech software and offer a customized piranha sales site as part of our product. Most of these are pretty straightforward sales websites, we generally do any heavy lifting via iframes or our other applications. Once we got customers over the "it's not wordpress" hump, they love Piranha, especially how quick and easy it is for us to spin up custom blocks/pages to fit whatever they need. 

One of the main reasons we went with piranha is that compared to other .net CMSs it is very straightforward to work with, and doesn't get in the way of us integrating our own libraries and custom elements. I also pushed for using it as I have been using it for various projects for years, it's genuinely my favorite CMS.

The main request I've had from my end users is that they want the ability to duplicate blocks. The other issue that crops up is just the overall performance of it, which admittedly is partly due to my implementation, but it would be useful to have some good guidance on optimizing piranha in the documentation.

From a developer perspective I'd love a way to easily backup the pages/blocks etc. from within the CMS and then re-upload a backup file to the CMS to hydrate the database. The main use cases I have for this are migrating between database types (e.g. going to SQLlite to SQL) and to make it easier for my dev team to grab live data to work locally.

I'd also find it useful to have the documentation include a guide for upgrading from one major version to the next, the 9.X to 10.X migration was a bit of a headache as we had to trial and error our way through the update order for the NuGet packages, and we had to poke around the source a bit to figure out what changes we needed to make to fix our custom middleware (although some of that was likely me just being dumb).

**Comment by jessemlay** [comment]
I use Piranha CMS in a heavily modified commercial app. One thing I'm currently looking at is moving away from gulp and integrating the build system with Vite mainly for HMR. I'm also curious as to how the move to .net 6 with the introduction of HMR for it will improve the developer experience.

**Comment by poptufty** [comment]
Thank you for the wonderful product.  There is support for lots of well thought out features that we can use.  We are looking at using Piranha in a multi-tenant environment, where each user will have their own set of pages to edit and support.  It looks like we might be able to use hooks for this, but we were wondering how close the development team is to providing support for this kind of usage.

**Reply by tidyui** [reply]
As we wrote in a previous answer, there’s no plan in turning Piranha into a true multi-tenant framework as it is in essence single-tenant. We are however thinking about adding hooks for when data for the different view models for the manager UI is constructed which would allow for filtering the visible data based on the current user. 

Regards

**Comment by BjarniPeturFridjonsson** [comment]
I use it for smaller companies and communities, I have used allot of different CMS systems and I like this one best of its no nonsense architecture, simplicity and extensibility.  Having a Headless CMS like this WITH proficient templates supercharged me and was the deciding factor in moving to Piranha. 

For improvements based on my experience is:
The admin pages around archives/news does not scale well when you have imported 22 years of community activity.
The User management becomes clunky when you have many users. 
The admin of image carousels/image galleries in pages is very bad.
For any improvements would probably be simplest solvable with search in admin and maybe grouping functionality but please continue the simplistic/clean methodology you have applied with great success here.

Onboarding/importing
bulk import of users, images(with metadata) and archives/news is missing, simple command line tools would be just awesome. I use direct injection into the databases to solve this, but is error prone.

Keep on the good work

**Comment by andyjwwhite** [comment]
Hi, firstly thanks for all your great work - nothing else has come close to this in terms of extensibility, clean code, attention to detail and carefully thought out features. We use Piranha at Sound in Theory to build highly customised bespoke websites for our clients. We often add extra views to the manager and work with data from our own EF contexts in the same DB, whilst using Piranha to manage auth, pages, content etc.

Here are my general thoughts:

- Scalability - Agree with @BjarniPeturFridjonsson that this is an issue in pages and content lists when you have larger data sets. Grouping / filtering would definitely help this and considering what to do when there are 50+ siblings in the page tree (eg. the ability to manage a list of routed content outside of the tree?)
- Manager JS - Consider moving to Vite for Manager tooling and refactoring to use Vue 3, es modules, dynamic imports etc.
- Permissions - per-item granular permissions. We've been asked about this a few times and we're pondering about how to implement it.
- Extensibility - it would be good to expand upon Piranha's great hooks (and config) to improve developer experience. One particular area is in image management - we built a module to add image cropping which presented some challenges.

Thanks again,

**Comment by tidyui** [comment]
Hi everybody! Let me take some time to address some of your awesome input:

**Scalability** [@andyjwwhite, @BjarniPeturFridjonsson]
I think this is a valid case. The archive UI was designed for smaller sets of data, and also for utilizing the build in system for injecting custom views into tabs for a page type. To provide a better more scalable solution we should probably look into a completely different solution for handling posts and post archive that lends itself better to large datasets.

**Manager JS** [@andyjwwhite, @jessemlay]
The manager UI (and its build process is getting a bit out of date). We've done some minor investigations into Vite and Vue 3 and we need to do an overhaul of the setup. As our current build process is a bit black magic and unorthodox I suggest this would require some redesign which would most likely create a lot of breaking changes for custom manager views.

**Extensibility** [@andyjwwhite]
Let's dive into the details of this as it's easier to understand if the requirements are more concrete.

**Duplicate blocks** [@EddyCwdry]
This is probably a pretty minor feature, I suggest that you could open an issue for this!

**Backup/Import/Export/Onboarding** [@BjarniPeturFridjonsson, @EddyCwdry]
I think we have an issue for this somewhere but unfortunately it's never been prioritized. It's however completely different features to implement a Piranha specific import/export and to implement a generic import format that can be used to import data from completely different systems.

**Image gallery** [@BjarniPeturFridjonsson]
Please open an issue with specifics with that doesn't work well :) Please note that the editing for this isn't anything build for the image gallery in particular, it's just a generic editing for a block group (which has three modes, master/detail, horizontal, vertical).

**Comment by GerardoGala** [comment]


**Reply by CarlSargunar** [reply]
Interesting that you don't like Umbraco, perfectly valid of course. Just curious since at its core I found it one of the simplest cms systems to use, particularly with modern versions of it. 

I think both these cms, Umbraco and piranha occupy very different spaces anyway, each has its own strengths and weaknesses.

**Reply by GerardoGala** [reply]
1. Umbraco isn't modular. I like a feature of package A and some features of package B. But I can not merge the two packages together.
2. I spent a week just to merge Contoso University Entity Framework sample web into Umbraco. And in the end the web site was slow. I only spent half a day t merge Contoso with Piranha.

Not a lot of samples for Piranha so learning it could be an uphill battle.

Thanks!

**Reply by CarlSargunar** [reply]
Yeah, Umbraco has definitely got issues in terms of customising it's admin interface, but in terms of it being modular, it was never meant to be modulare in that way - that you can just take a subset of features from a package. 

there are official and third party packages, which you can install, and a growing marketplace, but these are intended to be a full featured package, which you take all of. 

I'm curious to try Piranha though for customising it - if you say integrating a new feature set was easy that's a big plus

**Comment by GerardoGala** [comment]
Instead of asking us for features to improve Piranha, I think I have a better idea.

How about you build a fully functional web site to compete with [https://theclubspot.com/](url) ?

1. Add all the features in Piranha that are needed to build the best web site in the world for regatta, bike race, walkathon, and marathon management.
2. Make it free until you capture the market.
3. Create a for-profit company and transfer ownership of this website to this company. Then charge a competetive rate. This company will sponsor Piranha.org and now you have salary.

I can contribute by writing the Software Requirements Specification. I know how to manage sailing regatta and I can test the heck out of it.

You kill two birds with one stone. You will have a financial sponsor and this web site will propel Piranha to be the best free CMS.

Thanks!
